### PR TITLE
V1.2: Add functionality to edit person by name.

### PR DESCRIPTION
**Pull request by zioul123**
Using the format `edit [NameIdentifiers] [n/Name] [p/Phone] [e/Email] [a/Address] [t/tags]`, you can edit a user by name, as long as: NameIdentifiers when used with the find command (i.e. `find [NameIdentifiers]`) will return a single person in the list. If multiple people are found, or if no people are found, a command error is thrown.

Todo in v1.3: 
- System Test cases
- Add this extension to schedule, picture and delete commands

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/EditByNameCommand.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+    public EditByNameCommand(String personIdentifier, EditPersonDescriptor editPersonDescriptor) {
+        super(Index.fromZeroBased(NO_INDEX), editPersonDescriptor);
+        requireNonNull(personIdentifier);
+
+        this.personIdentifier = personIdentifier;
+    }
+
+    //@@author zioul123-reused
+    //Adapted from the execute method of EditCommand.
```
Comment (A19Sean) [inline src/main/java/seedu/address/logic/commands/EditByNameCommand.java]: Don't have to give credit to using AB4 code (nus-cs2103-AY1819S1/forum#90 (comment))

File: src/main/java/seedu/address/logic/parser/EditCommandParser.java [hunk 2]
```diff
@@ -79,4 +91,22 @@ public EditCommand parse(String args) throws ParseException {
...
 
+    //@@author zioul123
+    /**
+     * Attempts to parse an index from the provided preamble to a command.
+     * @param preamble The preamble of the command typed.
+     * @return An Optional of the index if the specified index is valid, an empty optional otherwise
+     */
+    private Optional<Index> getIndex(String preamble) {
+        //@@author zioul123-reused
+        //Moved from the parse method of EditCommandParser.
```
Comment (A19Sean) [inline src/main/java/seedu/address/logic/parser/EditCommandParser.java]: Same here

File: src/main/java/seedu/address/model/person/NameContainsAllKeywordsPredicate.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
+package seedu.address.model.person;
+
+import java.util.List;
+import java.util.function.Predicate;
+
+import seedu.address.commons.util.StringUtil;
+
+//@@author zioul123-reused
+//Based on NameContainsKeywordsPredicate, modified to test for all keywords instead of any keyword.
```
Comment (A19Sean) [inline src/main/java/seedu/address/model/person/NameContainsAllKeywordsPredicate.java]: Same here

File: src/test/java/seedu/address/logic/commands/EditByNameCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+import seedu.address.model.AddressBook;
+import seedu.address.model.Model;
+import seedu.address.model.ModelManager;
+import seedu.address.model.UserPrefs;
+import seedu.address.model.person.Person;
+import seedu.address.testutil.EditPersonDescriptorBuilder;
+import seedu.address.testutil.PersonBuilder;
+
+//@@author zioul123-reused
+//Based heavily on EditCommandTest, modified for EditByNameCommand
```
Comment (A19Sean) [inline src/test/java/seedu/address/logic/commands/EditByNameCommandTest.java]: Same here

File: docs/UserGuide.adoc [hunk 5]
```diff
@@ -109,31 +109,20 @@ Edits the name of the 2nd person to be `Betsy Crower` and clears all existing ta
 ==== Editing by name
 
 Edits an existing person in the address book by the person's name. +
-Format: `edit n/NAME [p/phone] [e/EMAIL] [a/ADDRESS] [t/TAG]...`
+Format: `edit IDENTIFIER [n/NAME] [p/phone] [e/EMAIL] [a/ADDRESS] [t/TAG]...`
```
[the remark this answers is not in the document]
Comment (zioul123) [reply docs/UserGuide.adoc]: Have made the change to use INDEX/EXISTING_NAME instead

**Pull request merged**